blockchain/chain.go

```diff
@@ -1682,6 +1693,173 @@ func (b *BlockChain) FetchHeader(hash *chainhash.Hash) (wire.BlockHeader, error)
 	return *header, nil
 }
 
+// locateInventory returns the node of the block after the first known block in
+// the locator along with the number of subsequent nodes needed to either reach
+// the provided stop hash or the provided max number of entries.
+//
+// In addition, there are two special cases:
+//
+// - When no locators are provided, the stop hash is treated as a request for
+//   that block, so it will either return the node associated with the stop hash
+//   if it is known, or nil if it is unknown
+// - When locators are provided, but none of them are known, nodes starting
+//   after the genesis block will be returned
+//
+// This is primarily a helper function for the locateBlocks and locateHeaders
+// functions.
+//
+// This function MUST be called with the chain state lock held (for reads).
+func (b *BlockChain) locateInventory(locator BlockLocator, hashStop *chainhash.Hash, maxEntries uint32) (*blockNode, uint32) {
+	// There are no block locators so a specific block is being requested
+	// as identified by the stop hash.
+	stopNode := b.index.LookupNode(hashStop)
+	if len(locator) == 0 {
+		if stopNode == nil {
+			// No blocks with the stop hash were found so there is
+			// nothing to do.
+			return nil, 0
+		}
+		return stopNode, 1
+	}
+
+	// Find the most recent locator block hash in the main chain.  In the
+	// case none of the hashes in the locator are in the main chain, fall
+	// back to the genesis block.
+	b.heightLock.RLock()
+	startNode := b.mainNodesByHeight[0]
+	b.heightLock.RUnlock()
+	for _, hash := range locator {
+		node := b.index.LookupNode(hash)
+		if node != nil && node.inMainChain {
+			startNode = node
+			break
+		}
+	}
+
+	// Start at the block after the most recently known block.  When there
+	// is no next block it means the most recently known block is the tip of
+	// the best chain, so there is nothing more to do.
+	if startNode != nil {
+		b.heightLock.RLock()
+		if startNode.height+1 < int64(len(b.mainNodesByHeight)) {
```

Comment:
This makes me wonder whether mainNodesByHeight should be a slice rather than a map, where the index in the slice is the node's block height.

Reply:
Good observation.  The eventual chain view will in fact use a slice, but for now this is a map because it will simplify the transition.  That said, there really is no need for this length check here.  It should just be setting `startNode` to the result of the lookup, which will either by the next node or nil as desired.

Changed.